### PR TITLE
refactor infrastructure and add ratelimit tests.

### PR DESCRIPTION
**Pull request by qicz**
- separate infrastructure configmap, service, and service account logic to proxy_xx and ratelimit_xx modules for easy maintenance. #1008 
- support #941

### CODE REVIEW

**Comment by arkodg** [comment]
hey @qicz thanks for looking into this, the refactor sgtm, would reccomend also renaming existing files into `proxy_X.go`
lastly a good way to test out your changes would be to get the tests for these methods in, avoiding any regression. I plan on working on https://github.com/envoyproxy/gateway/issues/941 soon but haven't started yet, is this something you can pick up ?

**Comment by qicz** [comment]
> hey @qicz thanks for looking into this, the refactor sgtm, would reccomend also renaming existing files into `proxy_X.go` lastly a good way to test out your changes would be to get the tests for these methods in, avoiding any regression. I plan on working on #941 soon but haven't started yet, is this something you can pick up ?

Let me try to do it

**Comment by Xunzhuo** [comment]
Assigned to you @qicz, thanks!

**Comment by qicz** [comment]
@arkodg I did it pls review this.

**Comment by Xunzhuo** [comment]
Great job @qicz, I will take a look ASAP.

File: internal/infrastructure/kubernetes/ratelimit_serviceaccount_test.go [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			kube := &Infra{
+				Client:    fakeclient.NewClientBuilder().WithScheme(envoygateway.GetScheme()).Build(),
+				Namespace: "test",
+			}
+			rateLimitInfra := new(ir.RateLimitInfra)
+			err := kube.deleteRateLimitServiceAccount(context.Background(), rateLimitInfra)
```
Comment (arkodg) [inline internal/infrastructure/kubernetes/ratelimit_serviceaccount_test.go:147]: I dont see `expect` being used , and should there be another test to check if the SA resource was actually deleted from the kube cluster ?

File: internal/infrastructure/kubernetes/ratelimit_serviceaccount_test.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+)
+
+func TestExpectedRateLimitServiceAccount(t *testing.T) {
+	cfg, err := config.New()
+	require.NoError(t, err)
+	cli := fakeclient.NewClientBuilder().WithScheme(envoygateway.GetScheme()).WithObjects().Build()
+	kube := NewInfra(cli, cfg)
+	infra := ir.NewInfra()
+
+	// An infra without Gateway owner labels should trigger
```
Comment (arkodg) [inline internal/infrastructure/kubernetes/ratelimit_serviceaccount_test.go:35]: can you raise a GH issue for this ?

File: internal/infrastructure/kubernetes/ratelimit_serviceaccount_test.go [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+	// an error.
+	_, err = kube.expectedProxyServiceAccount(infra)
```
Comment (arkodg) [inline internal/infrastructure/kubernetes/ratelimit_serviceaccount_test.go:37]: shouldnt this be expectedRateLimitX :) ?

File: internal/infrastructure/kubernetes/ratelimit_labels_test.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+func TestRateLimitLabels(t *testing.T) {
+	cases := []struct {
+		name     string
+		in       map[string]string
+		expected map[string]string
+	}{
+		{
+			name: "ratelimit-labels",
+			in:   map[string]string{"app.gateway.envoyproxy.io/name": rateLimitInfraName},
```
Comment (arkodg) [inline internal/infrastructure/kubernetes/ratelimit_labels_test.go:22]: why is `in` needed ?

File: internal/infrastructure/kubernetes/proxy_deployment.go [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+const (
+	// envoyContainerName is the name of the Envoy container.
+	envoyContainerName = "envoy"
+	// envoyNsEnvVar is the name of the Envoy Gateway namespace environment variable.
+	envoyNsEnvVar = "ENVOY_GATEWAY_NAMESPACE"
+	// envoyPodEnvVar is the name of the Envoy pod name environment variable.
+	envoyPodEnvVar = "ENVOY_POD_NAME"
+	// envoyCfgFileName is the name of the Envoy configuration file.
+	envoyCfgFileName = "bootstrap.yaml"
+	// envoyHTTPPort is the container port number of Envoy's HTTP endpoint.
```
Comment (arkodg) [inline internal/infrastructure/kubernetes/proxy_deployment.go:38]: @danehans is `envoyHTTPPort` and `envoyHTTPSPort` required ? doesnt the gateway API layer decide the infra ports ?

File: internal/infrastructure/kubernetes/ratelimit_service_test.go [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			kube := &Infra{
+				Client:    fakeclient.NewClientBuilder().WithScheme(envoygateway.GetScheme()).Build(),
+				Namespace: "test",
+			}
+			rateLimitInfra := new(ir.RateLimitInfra)
+			err := kube.deleteRateLimitService(context.Background(), rateLimitInfra)
```
Comment (arkodg) [inline internal/infrastructure/kubernetes/ratelimit_service_test.go:56]: shouldnt we insert the service into k8s cluster and check that the service has been deleted ?

**Comment by arkodg** [comment]
hey @qicz thanks for picking up #941 , left comments

File: internal/infrastructure/kubernetes/ratelimit_serviceaccount_test.go (hunk 3 above)
Comment (qicz) [reply internal/infrastructure/kubernetes/ratelimit_serviceaccount_test.go:37]: fixed

File: internal/infrastructure/kubernetes/ratelimit_serviceaccount_test.go (hunk 1 above)
Comment (qicz) [reply internal/infrastructure/kubernetes/ratelimit_serviceaccount_test.go:147]: I will remove the `except`. and just this one place to check sa delete.

File: internal/infrastructure/kubernetes/ratelimit_labels_test.go (hunk 4 above)
Comment (qicz) [reply internal/infrastructure/kubernetes/ratelimit_labels_test.go:22]: removed

File: internal/infrastructure/kubernetes/ratelimit_service_test.go (hunk 6 above)
Comment (qicz) [reply internal/infrastructure/kubernetes/ratelimit_service_test.go:56]: ok， I fixed this

File: internal/infrastructure/kubernetes/ratelimit_serviceaccount_test.go (hunk 2 above)
Comment (qicz) [reply internal/infrastructure/kubernetes/ratelimit_serviceaccount_test.go:34]: did it #1008

File: internal/infrastructure/kubernetes/ratelimit_service_test.go (hunk 6 above)
Comment (arkodg) [reply internal/infrastructure/kubernetes/ratelimit_service_test.go:56]: dont see the changes

Comment (qicz) [reply internal/infrastructure/kubernetes/ratelimit_service_test.go]: change again

**Review by arkodg — APPROVED** [review]
LGTM thanks @qicz !

File: internal/infrastructure/kubernetes/proxy_deployment.go (hunk 5 above)
Comment (arkodg) [reply internal/infrastructure/kubernetes/proxy_deployment.go:38]: thanks for analyzing, will raise a PR to clean this up

**Pull request merged**